src/connectionController.ts

```diff
@@ -116,142 +122,188 @@ export default class ConnectionController {
-      throw new Error(
-        'The connectionModel object is missing in saved connection info.'
+  async loadSavedConnections(): Promise<void> {
+    const globalAndWorkspaceConnections = {
+      ...this._storageController.get(
+        StorageVariables.GLOBAL_SAVED_CONNECTIONS,
+        StorageLocation.GLOBAL
+      ),
+      ...this._storageController.get(
+        StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
+        StorageLocation.WORKSPACE
+      ),
+    };
+
+    let connectionsDidChange = false;
+    const connectionsWithSecretsInKeytar: Array<{
+      connectionId: string;
+      connectionName: string;
+    }> = [];
+    const totalConnectionEntries = Object.entries(
+      globalAndWorkspaceConnections
+    );
+    for (const [connectionId, connectionInfo] of totalConnectionEntries) {
+      const connectionInfoWithSecret = await this._getConnectionInfoWithSecrets(
+        connectionInfo
+      );
+
+      if (connectionInfoWithSecret) {
+        connectionsDidChange = true;
+        this._connections[connectionId] = connectionInfoWithSecret;
+
+        if (
+          connectionInfoWithSecret.secretStorageLocation ===
+          SecretStorageLocation.Keytar
+        ) {
+          connectionsWithSecretsInKeytar.push({
+            connectionId,
+            connectionName: connectionInfo.name,
+          });
+        }
+      }
+    }
+
+    if (connectionsDidChange) {
+      this.eventEmitter.emit(DataServiceEventTypes.CONNECTIONS_DID_CHANGE);
+    }
+
+    if (connectionsWithSecretsInKeytar.length) {
+      log.error(
+        'Could not migrate secrets for a few connections',
+        connectionsWithSecretsInKeytar
+      );
+      this._telemetryService.track(
+        TelemetryEventTypes.KEYTAR_SECRETS_MIGRATION_FAILED,
+        {
+          totalConnections: totalConnectionEntries.length,
+          connectionsWithSecretsInKeytar: connectionsWithSecretsInKeytar.length,
+        }
+      );
+      void vscode.window.showInformationMessage(
+        [
+          'Could not migrate secrets for a few connections. Please review the following connections:',
+          connectionsWithSecretsInKeytar
+            .map(({ connectionName }) => connectionName)
+            .join(', '),
+        ].join('\n')
       );
     }
+  }
+
+  async _getConnectionInfoWithSecrets(
+    connectionInfo: StoreConnectionInfo
+  ): Promise<MigratedStoreConnectionInfoWithConnectionOptions | undefined> {
+    try {
+      if (connectionInfo.connectionModel) {
+        return this._migratePreviouslySavedConnection(
+          connectionInfo as StoreConnectionInfoWithConnectionModel
+        );
+      }
+
+      if (!connectionInfo.secretStorageLocation) {
+        return this._migrateKeytarSecrets(
+          connectionInfo as StoreConnectionInfoWithConnectionOptions
+        );
+      }
 
+      // We tried migrating this connection earlier but failed because Keytar was not
+      // available. So we return simply the connection without secrets.
```

Comment:
Could this happen when the user clicks away a request to unlock their OS keyring because they are repetitive/annoying if they have multiple connections stored? In that case it would be reasonable to assume that a subsequent attempt to get credentials would succeed, right?

Reply:
Yea, for such connections it would be safe to assume that they can be re-tried for migration.

There is one thing not so much clear to me - how / when would a user get a prompt to unlock the keyring? Asking this because it did not happen for me? Could be because I have been using the extension for sometime and might have allowed it at some point or could it be because `keytar` module does not prompt?

Reply:
> could it be because `keytar` module does not prompt?

`keytar` doesn’t do the prompting as far as I know, it integrates with the OS keychains instead, so when exactly the prompt occurs will depend on the OS and its configuration.

(I think we get prompts on macOS when upgrading Electron in Compass, for example. I’m not sure if something similar happens with VSCode at some point.)

Reply:
It happened before for vscode as well but does not happen anymore for a year or so. Not sure if we did something or vscode handles it differently. @Anemy do you have any context about it?

Reply:
In that case - I would assume that the `keytar` module will still be available (unless it was removed) but the `getPassword` call should fail which should throw an error. For this reason we will never mark such connection as migrated and thus they should be good for re-trying. But I should to test that.

Reply:
There's only a prompt for the first time the credential store is to be used for an app (only for some os's that are more stringent with permissions). If something with how the app is built (maybe bundle name? don't remember off the top of my head) changes then this also happens. When we update the electron version in compass-dev we see the prompts as the identifier changes then. As Alena mentioned it happened once around two years ago for VSCode which caused users to see this prompt again. I don't think we need to worry much about that, I don't see it as a bad thing to account for though.
As Anna said, the prompt is from the OS's credential store. _Maybe_ the user could have configured it in a way where they get a prompt each time it's requested.

https://github.com/mongodb-js/compass/blob/main/packages/hadron-build/lib/target.js

Reply:
Understood and thank you for the explanation here @Anemy 
In that case, we can still safely assume that keytar will be available to our extension, it is just that the calls to get secret would fail (if the user did not unlock or took too long to unlock), right?
If so then I would particularly account for that failure (originating from `keytar.getPassword`).

As of now, these errors would just bubble up and eventually get caught in `_getConnectionInfoWithSecret` and the connection will not be marked as migrated (ready to be re-attempted for migration).

Reply:
SGTM

Reply:
I just addressed this in my [commit here](https://github.com/mongodb-js/vscode/pull/552/commits/4250dd6a6a984fff425c50701aa8ea67c4c4286e). The connections which will fail migration for anything other than keytar not available are not marked as migrated and hence can be re-attempted again in the next load.

There is a kinda bad side effect of this which is - such connections (for which our `_migrateConnectionWithKeytarSecrets` rejected) will not appear in our Connection tree when the extension loads. I was also able to confirm from the [following code](https://github.com/mongodb-js/vscode/blob/5a9a6b4ac305ddb977e1371b33e13a822b6d16dc/src/connectionController.ts#L178-L208) that this is an existing behaviour.

I have created [this ticket](https://jira.mongodb.org/browse/VSCODE-443) to define and agree upon a behaviour of "loading of the connections" when we are not able to fetch the secrets (for whichever reason) for some of the connections. Please let me know if this is something we should absolutely fix within this PR.